core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -760,16 +899,27 @@ class ToBeCheckedRow(val row: TiRow,
                      val indexKeys: List[KeyWithDupInfo])
     extends Serializable {
 
-  //  val idxId: Long
-  // if handle is conflict or any key in indices key conflict with each other
-  // then two row equals
+  def isDuplicated: Boolean = {
+    handleKey.oldHandle != -1 || indexKeys.exists(key => key.oldHandle != -1)
+  }
+
+  // When the following case happens, two ToBeCheckedRows are considered as same:
+  // 1. handle key is same if not null
+  // 2. at least one of index keys is same and index keys is not empty
+  // 3. handle key is null and index keys are empty, but row is same.
   override def equals(that: Any): Boolean =
```

Comment:
consider the following case:

```
ToBeCheckRow1:
  handleKey = 1
  indexKeys = [11, 111]

ToBecheckRow2:
  handleKey = 1
  indexKeys = [22, 222]
```

`ToBeCheckRow1 == ToBecheckRow2`
but `ToBeCheckRow1.hashcode != ToBecheckRow2.hashcode`

Reply:
Yes. More refine is on the way